src/h5web/providers/hsds/utils.ts

```diff
@@ -26,3 +30,66 @@ export function assertHsdsDataset(
     throw new Error('Expected entity to be HSDS dataset');
   }
 }
+
+export function convertHsdsBaseType(hsdsBaseType: HsdsBaseType): HDF5BaseType {
+  if (hsdsBaseType.class === HDF5TypeClass.String) {
+    return hsdsBaseType;
+  }
+
+  const { class: hsdsClass, base } = hsdsBaseType;
+
+  const regex = /H5T_(?:STD|IEEE)_([A-Z])(\d+)(BE|LE)/u;
+  const matches = regex.exec(base);
+
+  if (!matches) {
+    throw new Error(`Unrecognized base ${base}`);
+  }
+
+  const [, sign, size, endianness] = matches;
+  assertEndianness(endianness);
+
+  return {
+    class: hsdsClass,
+    endianness,
+    size: Number.parseInt(size, 10),
+    unsigned: sign === 'U' || undefined,
+  };
+}
+
+export function convertHsdsType(hsdsType: HsdsType): HDF5Type {
+  if (isString(hsdsType)) {
+    return hsdsType;
+  }
+
+  switch (hsdsType.class) {
+    case HDF5TypeClass.Array:
+      return {
+        class: HDF5TypeClass.Array,
+        base: convertHsdsBaseType(hsdsType.base),
+        dims: hsdsType.dims,
+      };
+
+    case HDF5TypeClass.VLen:
+      return {
+        class: HDF5TypeClass.VLen,
+        base: convertHsdsBaseType(hsdsType.base),
+      };
```

Comment:
You can have two `case` in a row:

```
case HDF5TypeClass.Array:
case HDF5TypeClass.VLen:
  return { class: hsdsType.class, ... }
```

Reply:
There is a `dims` field for `HDF5TypeClass.Array` and not for `HDF5TypeClass.VLen`, right ?

Reply:
yeah, but you had them both in the same `if` earlier, didn't you?

Reply:
nevermind if it's clearer like this, though

Reply:
> yeah, but you had them both in the same `if` earlier, didn't you?

Yes because I was doing `...hsdsType` but I also find it much clearer like this.